### PR TITLE
Refactor reusable code and add BaseBatchDataTask

### PR DESCRIPTION
**Pull request by prescod**
A different PR will pull in the FactoryBoy stuff.

### CODE REVIEW

**Comment by prescod** [comment]
The last check is hung up on some super-slow SF Org but I'm assuming it will turn out fine.

File: cumulusci/tasks/bulkdata/base_generate_data_task.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+from sqlalchemy.ext.automap import automap_base
+
+from cumulusci.core.tasks import BaseTask
+from cumulusci.core.utils import ordered_yaml_load
+
+from .utils import create_table
+
+
+class BaseGenerateDataTask(BaseTask):
+    """Abstract base class for any class that generates data using a SQL DB."""
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/base_generate_data_task.py]:
```suggestion
    """Abstract base class for any class that generates data in a SQL DB."""
```

File: cumulusci/tasks/bulkdata/tests/test_base_generate_data_tasks.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                DummyBaseBatchDataTask,
+                {
+                    "options": {
+                        "num_records": NUM_RECORDS,
+                        "mapping": mapping_file,
+                        "database_url": dburl,
+                    }
+                },
+            )
+            task._testfilename = tmp_db_path
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/tests/test_base_generate_data_tasks.py]: Why's this needed? Can't the task check the database_url in its options?

File: cumulusci/tasks/bulkdata/tests/test_base_generate_data_tasks.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            task()
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/tests/test_base_generate_data_tasks.py]: I see the assertions in DummyBaseBatchDataTask.generate_data, but there's nothing asserting that it is actually called.

File: cumulusci/tasks/bulkdata/utils.py [hunk 4]
```diff
@@ -123,3 +127,48 @@ def _sql_bulk_insert_from_csv(self, conn, table, columns, data_file):
...
+    else:
+        fields.append(Column("id", Integer(), primary_key=True, autoincrement=True))
+
+
+def create_table(mapping, metadata):
+    """Given a mapping data structure (from mapping.yml) and SQLAlchemy
+       metadata, create a table matching the mapping.
+
+       Mapping should be a dict-like with keys "fields", "table" and
+       optionally "oid_as_pk" and "recored_type" """
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/utils.py]:
```suggestion
       optionally "oid_as_pk" and "record_type" """
```

File: cumulusci/tasks/bulkdata/tests/test_base_generate_data_tasks.py (hunk 3 above)
Comment (prescod) [reply cumulusci/tasks/bulkdata/tests/test_base_generate_data_tasks.py]: 9fb89c34116a8463da29f6a508ca018e24db152a

File: cumulusci/tasks/bulkdata/tests/test_base_generate_data_tasks.py (hunk 2 above)
Comment (prescod) [reply cumulusci/tasks/bulkdata/tests/test_base_generate_data_tasks.py]: 9fb89c34116a8463da29f6a508ca018e24db152a

**Pull request merged**